framework/di/Container.php

```diff
@@ -527,7 +531,17 @@ public function resolveCallableDependencies(callable $callback, $params = [])
                 } elseif (isset(Yii::$app) && Yii::$app->has($name) && ($obj = Yii::$app->get($name)) instanceof $className) {
                     $args[] = $obj;
                 } else {
-                    $args[] = $this->get($className);
+                    // If the argument is optional we catch not instantiable exceptions.
+                    try {
+                        $args[] = $this->get($className);
+                    } catch (NotInstantiableException $e) {
```

Comment:
Following statements contradict each other:

> no; we still want to be able to inject optional dependencies.

and 

> If it is optional then requiring configuration in DI doesn't make sense.

I am fine having better exception. But for me the proposed behavior change looks inconsistent.
It seems we should skip injection for any param with default value or attempt to instantiate any with type specified.

Silently ignoring just interfaces does not seem right.